### PR TITLE
[SPARK-30620][SQL] avoid unnecessary serialization in AggregateExpression

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

Expressions are very likely to be serialized and sent to executors, we should avoid unnecessary serialization overhead as much as we can.

This PR fixes `AggregateExpression`.

### Why are the changes needed?

small improvement

### Does this PR introduce any user-facing change?

no

### How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @maropu

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala [hunk 1]
```diff
@@ -108,9 +109,10 @@ case class AggregateExpression(
     UnresolvedAttribute(aggregateFunction.toString)
   }
 
-  lazy val filterAttributes: AttributeSet = filter.map(_.references).getOrElse(AttributeSet.empty)
+  def filterAttributes: AttributeSet = filter.map(_.references).getOrElse(AttributeSet.empty)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala]: I checked and we don't call it repeatedly. `def` is good enough.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala [hunk 2]
```diff
@@ -96,6 +96,7 @@ case class AggregateExpression(
   extends Expression
   with Unevaluable {
 
+  @transient
   lazy val resultAttribute: Attribute = if (aggregateFunction.resolved) {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala]: Oh, I checked this method is called inside the planning phase (`AggUtils` and `PhysicalAggregation`) only, so this fix looks reasonable.

**Review by maropu — APPROVED** [review]

**Comment by cloud-fan** [comment]
The R test is broken now, and is not related to this PR.

Thanks for the review, merging to master!

**Comment by gengliangwang** [comment]
Late LGTM :)

**Pull request closed without merging**